addons/stock/models/stock_quant.py

```diff
@@ -221,17 +215,125 @@ def action_view_stock_moves(self):
     @api.model
     def action_view_quants(self):
         self = self.with_context(search_default_internal_loc=1)
-        if not self.user_has_groups('stock.group_stock_multi_locations'):
-            company_user = self.env.company
-            warehouse = self.env['stock.warehouse'].search([('company_id', '=', company_user.id)], limit=1)
-            if warehouse:
-                self = self.with_context(default_location_id=warehouse.lot_stock_id.id)
-
-        # If user have rights to write on quant, we set quants in inventory mode.
-        if self.user_has_groups('stock.group_stock_manager'):
-            self = self.with_context(inventory_mode=True)
+        self = self._set_view_context()
         return self._get_quants_action(extend=True)
 
+    @api.model
+    def action_view_inventory(self):
+        """ Similar to _get_quants_action except specific for inventory adjustments (i.e. inventory counts). """
+        self = self._set_view_context()
+        self._quant_tasks()
+
+        ctx = dict(self.env.context or {})
+        ctx['no_at_date'] = True
+        action = {
+            'name': _('Stock On Hand'),
+            'view_type': 'tree',
+            'view_mode': 'list',
+            'view_id': self.env.ref('stock.view_stock_quant_tree_inventory_editable').id,
+            'res_model': 'stock.quant',
+            'type': 'ir.actions.act_window',
+            'context': ctx,
+            'domain': [('location_id.usage', 'in', ['internal', 'transit'])],
+            'help': """
+                <p class="o_view_nocontent_barcode_scanner">
+                    Want to speed up your inventory counts? Try our Barcode app
+                </p><p>
+                    Barcode scanner can be activated via inventory settings.
+                    Manual inventory adjustments can also be performed and pre-filled with
+                    suggested counted quantity.
+                </p>
+                """
+        }
+        return action
+
+    def action_apply_inventory(self):
+        quants_outdated = self.env['stock.quant']
+        quants_tracked_without_lot = self.env['stock.quant']
+        quants_duplicate_sn = self.env['stock.quant']
+        quants_to_process = self.env['stock.quant']
+        for quant in self:
+            rounding = quant.product_uom_id.rounding
+            if float_compare(quant.inventory_diff_quantity, 0.0, precision_rounding=rounding) == 0\
+                and float_compare(quant.inventory_quantity, 0.0, precision_rounding=rounding) == 0\
+                and float_compare(quant.quantity, 0.0, precision_rounding=rounding) == 0:
+                continue
+            if (quant.inventory_quantity - quant.inventory_diff_quantity) != quant.quantity:
+                quants_outdated |= self
```

Comment:
Perf remark, |= can be slow when len(self) increase. Moreover, you get only the ids at the end (except quants_to_process). You can use sets and add() the quant.id in the ifs

Reply:
Done. Used list so I don't have to convert the set to a list. I'm also not sure why there was a `quants_to_process`so that's deleted as well as the unnecessary `quants_duplicate_sn` (there's still a warning when users create a duplicated SN anyways, we're still missing a way to highlight when there already existing duplicate SNs....).